### PR TITLE
Add new Checkbox component

### LINKED ISSUE #1430: Create a checkbox component
Raised by @johnHackworth 

Right now, most of the checkboxes in the platform are just regular `<input type="checkboxes">`, styled to use a noticon on their :before psedoselector to change how they look when checked. 

We should create a new react component that encapsulates the functionality and use `Gridicon` for it, avoiding CSS hacks to customize the look. 

It's worth noting that we already have https://github.com/Automattic/wp-calypso/tree/master/client/components/forms/form-checkbox . Maybe it would be better to just rename that component (since it's not doing anything form-related, it could become `client/components/checkbox` ) and work from it. And if in the future we need any checkbox with a form-related behavior, add a new `from-checkbox` component again.

### PR DESCRIPTION
**Pull request by nobuti**
This PR fixes #1430.

This is the base for a checkbox component. I've mimicked the default checkbox component, all the look and feel but the check mark, the gridicon is included here. The default size is 18, 12 is too small IMO. Besides the design part, I've created it as a new component under the `client/components` folder but I didn't clean up the old form control component as the issue description's mentioned. I wanted to assure the component is right before to get deeper.

### CODE REVIEW

**Comment by nobuti** [comment]
cc @johnHackworth

**Comment by johnHackworth** [comment]
pinging @rickybanister, @MichaelArestad & @mtias

**Comment by johnHackworth** [comment]
good one, thank you @nobuti !

File: client/components/checkbox/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Checkbox is a React component used to render a checkbox input field. It is essentially an enhanced version of `<input type="checkbox" />`, using Gridicon as check mark.
+
+## Usage
+
+Refer to the following code snippet for a typical usage example:
+
+```jsx
+<Checkbox
+	checked={ this.state.showSubtitles }
+	onChange={ this._toggleShowSubtitles }>
```
Comment (dmsnell) [inline client/components/checkbox/README.md]: Since [our style guide](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md) tells us to avoid prefixing a function name with `_`, we probably shouldn't use that in an Readme example :wink:

File: client/components/checkbox/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	Show subtitles
+</Checkbox>
+```
+
+## Props
+
+Props not listed below will be passed automatically to the rendered range input element.
```
Comment (dmsnell) [inline client/components/checkbox/README.md]: This is probably something that's widely-enough known and universal to React components that we probably don't need it here either. Leaving it in could confuse people who expect this behavior, thinking that the mere mention of it might indicate that something is _not_ normal about how the props get passed.

File: client/components/checkbox/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+### `checked` (`boolean`)
+
+A boolean indicating whether the checkbox should be checked or not, then the checkmark Gridicon is shown.
```
Comment (dmsnell) [inline client/components/checkbox/README.md]: We could skip the first few words here since they are already right above

```js
### `checked` (`boolean`)

Whether or not the checkbox should be checked.
```

File: client/components/checkbox/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+### `disabled` (`boolean`)
+
+A boolean indicating whether the checkbox should be disabled or not.
+
+### `onChange` (`function`)
+
+An optional function to communicate the state change.
```
Comment (dmsnell) [inline client/components/checkbox/README.md]: > An optional function **called** when the state changes

this is probably more appropriate as a description of what happens with that callback

File: client/components/checkbox/index.jsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import React from 'react';
+import classNames from 'classnames';
+import hasOwn from 'lodash/object/has';
+
+/**
+* Interal dependencies
+*/
+import Gridicon from 'components/gridicon';
+
+module.exports = React.createClass( {
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: Nice `import`s, how about some `export`s?

```js
export default React.createClass( {
	…
} );
```

File: client/components/checkbox/index.jsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	displayName: 'Checkbox',
+
+	propTypes: {
+		onChange: React.PropTypes.func,
+		checked: React.PropTypes.any
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: we actually specify a boolean here in the documentation, so we should match that here…

```js
import React, { PropTypes } from 'react';

propTypes: {
	checked: PropTypes.bool.isRequired,
	disabled: PropTypes.bool,
	onChange: PropTypes.func
}
```

File: client/components/checkbox/index.jsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	},
+
+	getDefaultProps() {
+		return {
+			checked: false,
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: If this _is_ required, then we don't need a default value

File: client/components/checkbox/index.jsx [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+		};
+	},
+
+	getInitialState() {
+		const props = this.props;
+		const checked = hasOwn( props, 'checked' ) ? props.checked : false;
+		return {
+			checked: checked
+		}
+	},
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: If we make `checked` a required prop, then we should have enough protection from the developer warnings in the console to guard against having an invalid value there, which would eliminate our need for `hasOwn()` and would simplify this function

```js
getInitialState() {
	const { checked } = this.props;

	return { checked };
}
```

File: client/components/checkbox/index.jsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			focused: false
+		} )
+	},
+
+	render() {
+		const focusCssClass = this.getFocusedCss();
+		const disabledCssClass = this.getDisabledCss();
+
+		return (
+			<label className={ classNames( this.props.className, 'checkbox', focusCssClass, disabledCssClass ) } >
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: `classNames` actually supports boolean conditionals already!

```js
const { checked, focused } = this.state;
const { children, disabled } = this.props;
const classes = classNames( 'checkbox', {
	'is-focused': focused,
	'is-disabled': disabled
} );
```

File: client/components/checkbox/index.jsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+				<input type="checkbox"
+					checked = { this.isChecked() }
+					disabled = { this.props.disabled }
+					onChange = { this.handleChange }
+					onFocus = { this.handleFocus }
+					onBlur = { this.handleBlur } />
+				{ this.props.children }
+				{ this.renderMark() }
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: We tend to prefer to render the components directly in `render()` if they don't require really complicated logic to produce, which would probably be the case here, because it makes the component as a whole easier to see when reading

```js
render() {
	const { checked } = this.state;
	const { children } = this.props;

	return (
		<input type="checkbox" … />
		{ children }
		{ checked && <Gridicon icon="checkmark" size={ 18 } /> }
	);
}
```

File: client/components/checkbox/index.jsx (hunk 9 above)
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: if someone wants to add an additional CSS class to this component, they can actually do so when they use it

```jsx
<Checkbox className="fancyClass" />
```

which means we won't need to copy it here as well

File: client/components/checkbox/index.jsx [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+		return (
+			<label className={ classNames( this.props.className, 'checkbox', focusCssClass, disabledCssClass ) } >
+				<input type="checkbox"
+					checked = { this.isChecked() }
+					disabled = { this.props.disabled }
+					onChange = { this.handleChange }
+					onFocus = { this.handleFocus }
+					onBlur = { this.handleBlur } />
+				{ this.props.children }
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: I like nesting the text or whatever inside the component like you did here over setting a `value` prop like in the HTML specification. It might be worth pointing this out explicitly (that the implementations differ) in the Readme to warn others

**Comment by dmsnell** [comment]
@nobuti thanks for the contribution! I left you a few comments and I think they should get you some momentum here. if you want to call me back after you review them I would be happy to go over it all a second time. I didn't write down everything I could have, but I wanted to give you some ideas and let you think about them before adding more noise to the code diff.

**Comment by nobuti** [comment]
@dmsnell thanks for taking the time to code review. I'll take a deeper look and I'll ping you when it's ready for a second pass.

**Comment by nobuti** [comment]
@dmsnell ready for a second pass. All comments were took in considerations. Let me know what you think it can be improved.

**Comment by rickybanister** [comment]
It would be great if you would swap the checkbox examples in the devdocs/design ui components for this new component. That's a good first step in converting existing checkboxes to the new component (and preventing more non-componentized checkboxes from being created across the project).

**Comment by nobuti** [comment]
@rickybanister sure thing! I'll do as soon as possible.

**Comment by nobuti** [comment]
@rickybanister done. I've updated all occurrences but the bulk select component. I guess this is meat for another issue.

**Comment by dmsnell** [comment]
@nobuti I apologize as it might be a couple of days before I can get back to this. I'm more than happy to go over it again but at the moment I'm swamped. Thanks for the updates!

**Comment by nobuti** [comment]
@dmsnell I get it, no problem at all, ping me if you need anything.

File: client/components/checkbox/index.jsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	getDefaultProps() {
+		return {
+			disabled: false
+		};
+	},
+
+	handleChange( e ) {
+		const { onChange } = this.props;
+
+		onChange && onChange( e, this.state.checked );
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: This is a case where we have a bunch of code just to make sure we don't call a function that doesn't exist. If, however, we added a default `onChange` handler, we could know that we could always call this.

```js
import noop from 'lodash/utility/noop';

getInitialProps() {
	return {
		disabled: false,
		onChange: noop
	};
}

handleChange( e ) {
	this.props.onChange( e, this.state.checked );
}
```

File: client/components/checkbox/index.jsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+	handleBlur() {
+		this.setState( {
+			focused: false
+		} )
+	},
+
+	render() {
+		const { focused } = this.state;
+		const { className, children, checked, disabled, id, name } = this.props;
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: Oops - don't forget to put these (`id` and `name`) in the `propTypes` declaration and documentation! 😉

File: client/components/checkbox/style.scss [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	font-size: 14px;
+	margin-bottom: 5px;
+	position: relative;
+	user-select: none;
+}
+
+.checkbox input[type="checkbox"] {
+	@include hide-content-accessibly;
+	border: 0;
+	margin: -1px;
```
Comment (dmsnell) [inline client/components/checkbox/style.scss]: I'm just curious, not leaving a comment about the code, but why the `-1px` margin here?

File: client/components/checkbox/style.scss [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+	padding: 0;
+}
+
+.checkbox .gridicon {
+	fill: $blue-medium;
+	left: 10px; // half width of box plus 1px of border
+	position: absolute;
+	top: 50%;
+	transform: translate(-50%, -50%);
+	z-index: 1;
```
Comment (dmsnell) [inline client/components/checkbox/style.scss]: @drw158 or @folletto could one of you review this `z-index`? I can't remember for sure but I thought that we were making some kind of standardization for declaring these and this is a low-level application component.

File: client/components/checkbox/style.scss [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+	height: 18px;
+	left: 0;
+	position: absolute;
+	top: 50%;
+	transform: translateY(-50%);
+	width: 18px;
+}
+
+.checkbox:not(.is-disabled):hover:after {
+	border-color: #a8bece;
```
Comment (dmsnell) [inline client/components/checkbox/style.scss]: ping for designers - do we want/need to declare these colors somewhere as a variable?

File: client/components/forms/docs/example.jsx [hunk 17]
```diff
@@ -89,8 +98,7 @@ var FormFields = React.createClass( {
 							id="username"
 							name="username"
 							disabled
-							placeholder="Placeholder text..."
-						/>
+							placeholder="Placeholder text..." />
```
Comment (dmsnell) [inline client/components/forms/docs/example.jsx]: It looks like this and the following changes may have been introduced by an auto formatter? Could we remove them from the PR? Whenever we have formatting issues we like to separate them into their own PRs to make it easy to pinpoint changes with revisions. In this case, it's hard to tell if something changed because of actual logic changes or because we missed something in the formatting update.

File: client/components/checkbox/index.jsx [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+		return (
+			<label className={ classes } >
+				<input type="checkbox"
+					id = { id }
+					name = { name }
+					checked = { checked }
+					disabled = { disabled }
+					onChange = { this.handleChange }
+					onFocus = { this.handleFocus }
+					onBlur = { this.handleBlur } />
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: a quick note on formatting here - we don't include spaces around the equals signs

```js
id={ id }
name={ name }
…
```

File: client/components/checkbox/style.scss (hunk 14 above)
Comment (nobuti) [inline client/components/checkbox/style.scss]: TBH, I don't know, I copied the styles for the `componentes/forms/checkbox` and I presumed that it was for something, so I preferred to keep it.

File: client/components/checkbox/index.jsx [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+		this.setState( {
+			focused: true
+		} )
+	},
+
+	handleBlur() {
+		this.setState( {
+			focused: false
+		} )
+	},
```
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: concerning the change, blur, and focus functions - [we prefer to give descriptive names](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#react-method-names), so for instance, we could use these instead:

```js
toggleCheckmark( e ) {
	this.props.onChange( e, this.state.checked );
}

focus() {
	this.setState( { focused: true } );
}

blur() {
	this.setState( { focused: false } );
}

…

onChange={ this.toggleCheckmark }
onFocus={ this.focus }
onBlur={ this.blur }
```

File: client/components/checkbox/index.jsx (hunk 12 above)
Comment (nobuti) [inline client/components/checkbox/index.jsx]: Awesome! Let me fix it.

File: client/components/forms/docs/example.jsx (hunk 17 above)
Comment (nobuti) [inline client/components/forms/docs/example.jsx]: My fault, the linter complained about this, so I fixed. I'll revert them.

File: client/components/checkbox/index.jsx (hunk 12 above)
Comment (dmsnell) [inline client/components/checkbox/index.jsx]: `this.state.checked` is passed as a parameter here but doesn't exist. did you mean to send `this.props.checked`

actually, I think we could simplify this API to send only a single boolean value - the new checked status. that is, if we have a checkbox that is already checked, we should send out `false` when clicked to indicate the new status.

```js
toggleCheckmark() {
	const { checked, onChange } = this.props;

	onChange( ! checked );
}
```

**Comment by dmsnell** [comment]
This is really looking great @nobuti! :+1: 

Besides the comments I left in the code, I have a couple higher-level points to consider:

 - Because the responsibility for keeping track of the checked-state lies with the calling parent, it's probably worth mentioning in the Readme. I think that the way you have it is best and maybe everyone already knows to expect this, but do whatever you think is best.
 - Some of the variable names here ought to be updated to match [our style guide](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#variable-names). For example, `disabled` should be `isDisabled` to indicate that it's a binary state.
 - I tested this and it looks great. If you wanted to make it totally amazing, you could look into adding some tests for the component. You can refer to [a component I wrote](https://github.com/Automattic/wp-calypso/tree/master/client/lib/interval) and its tests for an example, though it may not be the simplest one. Tests need both the test directory and a copy of the `Makefile` in the component-root directory. Further, I found `enzyme` to be incredibly useful for testing React components.
    - Tests could verify things including but not limited to the following:
        - Does the Gridicon appear when `isChecked` is `true`
        - Does it not appear when `isChecked` is `false`
        - Do the individual CSS classes appear when appropriate given the input
        - Does the parent `onChange` event get called appropriately when clicked.

If you are willing to add tests and need some help, please feel free to ping me or anyone else. @aduth is a master at testing, just to name one.

This is awesome; please don't be discouraged by all my comments. Typically, the lower-level a component is, the more rigor we must apply to it before merging since it will be used in countless ways and places. Thanks for the earlier changes and happy hacking!

**Comment by nobuti** [comment]
@dmsnell thank you for and awesome an super useful review, I have some tests made but I need to read about your testing environment in order to run them. I'll try to integrate them in the next few days.

**Comment by nobuti** [comment]
Tests added!

**Comment by dmsnell** [comment]
stellar job @nobuti. those tests looks great.

I would still like to see `disabled` and `checked` become `isDisabled` and `isChecked` unless you think there's a valid reason to keep them as-is.

we still need to have some designers review this PR to look over the visual and CSS aspects, but this should be ready after that.

the last step we will do is squash all the commits into one. we normally do that to keep the main tree tidier, whereas one can see each new macro-level change as a single commit instead of a bunch of commits building up to the "final product."

**Comment by nobuti** [comment]
sorry @dmsnell, I missed that! It's fixed now. I'll wait until the design review to rebase the history.

**Comment by dmsnell** [comment]
:+1:

**Comment by rickybanister** [comment]
I just took a look at the branch and everything feels pretty close. 

I noticed a few small things:

* There is a slight visual regression—when the input is focused and hovered the non-focused hover state is triggered. The old checkbox css didn't have this issue. It feels a little funny to have the border flicker when focused and hovering. 
* I don't think we need the `border-radius: 1px;` It can only really be rendered on a retina screen, no? The old checkbox was simply a square and I think it's fine to keep it that way.
* This checkbox is 18px square and the old one was 16px. The radio button is also 16px. The gridicon shouldn't be used at 16, but perhaps we can manage an 18px gridicon within the 16px square? I might let another designer chime in here, but it would be nice to keep the check and radio the same size.

File: client/components/checkbox/style.scss (hunk 16 above)
Comment (drw158) [inline client/components/checkbox/style.scss]: Yes, all colors should be a variable, no hex codes.

@nobuti you can find the color variables here https://wordpress.com/design-handbook/colors/

or in the colors.scss file

**Comment by drw158** [comment]
> This checkbox is 18px square and the old one was 16px. The radio button is also 16px. The gridicon shouldn't be used at 16, but perhaps we can manage an 18px gridicon within the 16px square? I might let another designer chime in here, but it would be nice to keep the check and radio the same size.

An 18px icon within a 16px square could work, since the bounding box of the gridicon is larger than the icon itself. I would like to see it and test it out.

File: client/components/checkbox/style.scss (hunk 15 above)
Comment (drw158) [inline client/components/checkbox/style.scss]: It looks like the z-index is just to get the custom icon on top, not sure about any standardization efforts across Calypso.

Comment (nobuti) [inline client/components/checkbox/style.scss]: I've refactored the styles to use the :before pseudo-element to «draw» the border, so no more z-index.

File: client/components/checkbox/style.scss (hunk 16 above)
Comment (nobuti) [inline client/components/checkbox/style.scss]: :+1:

**Comment by nobuti** [comment]
@rickybanister @drw158 ready for another pass.

**Comment by MichaelArestad** [comment]
@drw158 Honestly, I was thinking a custom set of three icons could be used just for checkboxes. They don't have to be part of Gridicons.

**Comment by aduth** [comment]
>I would still like to see `disabled` and `checked` become `isDisabled` and `isChecked` unless you think there's a valid reason to keep them as-is.

While valid for normal variable usage, I've started to see this encouraged less for component props. An example would be the [`<Button />` component](https://github.com/Automattic/wp-calypso/blob/4d4e305417dbede48a07cc041e1f4434e24b4f9e/client/components/button/index.jsx#L14-L21). I think the idea is to align them with the "props" (truly properties) of baked-in DOM nodes, e.g. `<input type="checkbox" checked disabled readOnly />`.

File: client/components/forms/docs/example.jsx [hunk 20]
```diff
@@ -77,10 +82,14 @@ var FormFields = React.createClass( {
 
 					<FormFieldset>
 						<FormLegend>Form Checkbox</FormLegend>
-						<FormLabel>
-							<FormCheckbox id="comment_like_notification" name="comment_like_notification" />
-							<span>Email me when someone Likes one of my comments.</span>
-						</FormLabel>
+
+						<Checkbox
+							id="comment_like_notification"
```
Comment (aduth) [inline client/components/forms/docs/example.jsx]: Outside the scope of this pull request, but I wonder whether there's truly a need for an `id` here (or even `name`?).

I've seen some opposition to relying on `props` spread too much, as it obscures what's being passed, but it can be handy for allowing a component to pass along miscellaneous props, especially when it's directly wrapping a DOM node that accepts a variety of attributes (props). In other words, rather than adding [`indeterminate`](https://developer.mozilla.org/en-US/docs/Web/CSS/:indeterminate) (or in this case, `id` or `name`) as a prop of `<Checkbox />` in the rare case it's ever needed, rendering its `<input />` as:

```jsx
<input { ...this.props } type="checkbox" />
```

File: client/components/checkbox/test/index.jsx [hunk 21]
```diff
@@ -0,0 +1,56 @@
+require( 'lib/react-test-env-setup' )();
```
Comment (aduth) [inline client/components/checkbox/test/index.jsx]: Tests still pass without including this module. In particular, the `react-test-env-setup` module is handy when needing to emulate a true DOM, which seems to be handled here by React/`enzyme` exclusively using shallow rendering.

File: client/components/checkbox/test/index.jsx [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+		} );
+
+		it( 'should render disabled if true', function() {
+			const checkbox = shallow( <Checkbox isDisabled={ true } isChecked={ true } /> );
+			expect( checkbox.html() ).to.contain( 'disabled' );
+		} );
+	} );
+
+	describe( 'Events', function() {
+		it('should pass change events', function() {
```
Comment (aduth) [inline client/components/checkbox/test/index.jsx]: Nitpicking: This needs a space after the opening paren after `it(`

If your editor of choice supports it, you may be interested in finding a linting plugin supporting our [ESLint configuration](https://github.com/Automattic/wp-calypso/blob/master/.eslintrc), as it helps to spot these types of issues. There's a few examples [in our docs](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#eslint), including command-line usage.

File: client/components/checkbox/test/index.jsx (hunk 21 above)
Comment (nobuti) [inline client/components/checkbox/test/index.jsx]: :+1:

File: client/components/checkbox/test/index.jsx (hunk 22 above)
Comment (nobuti) [inline client/components/checkbox/test/index.jsx]: I missed that! Installed for my editor.

File: client/components/forms/docs/example.jsx (hunk 20 above)
Comment (nobuti) [inline client/components/forms/docs/example.jsx]: I see your point, but I tried to be explicit. You are right about the `id` but `name` attribute is used when sending data in a form submission. Considering this is a form field, I tend to include it always.

**Comment by nobuti** [comment]
@aduth @dmsnell let me know if you want me to revert the naming convention for the props.

File: client/components/forms/docs/example.jsx (hunk 20 above)
Comment (aduth) [inline client/components/forms/docs/example.jsx]: >I see your point, but I tried to be explicit.

Yep, no need to change anything here. Thinking out loud a bit about when or if it's ever appropriate to use the props spread :smile: 

Thanks for the information about the need for `name`.

**Comment by drw158** [comment]
@MichaelArestad That could work, I didn't think about that. What 3 icons would we need?

**Comment by MichaelArestad** [comment]
@drw158 a plus, a check, and a dash (for the tristate checkbox). I had done some super quick mockups when discussing the bulk select component initially with heavier weighting ( https://github.com/Automattic/wp-calypso/issues/141). 

image

I think making them a touch smaller might also improve them.

**Comment by drw158** [comment]
That sounds like a good idea to me.

**Comment by blowery** [comment]
We just landed lodash 4, which changes how we include packages from lodash. This PR includes a new import of lodash that needs to be fixed. See https://github.com/lodash/lodash/wiki/Changelog#compatibility-warnings for all of the changes.

**Comment by nobuti** [comment]
Also, I'm not sure if the scope of this issue has changed, according to this comment https://github.com/Automattic/wp-calypso/pull/2008#issuecomment-170725702. Right now the PR is built on Gridicon.

**Comment by lancewillett** [comment]
This one is stalled, see also #5323 which is smaller change and actionable now.

Closing this one, but feel free to re-open if it's actionable and can be pushed in.
